### PR TITLE
Refactor logging, in particular decoupling the parser and logging.

### LINKED ISSUE #331: Refactor logging across sqlfluff
Logging within the rules engine has had a recent upgrade to leverage the python logging library properly. The parser has not had this treatment and is inefficient and inflexible as a result. This issue is to refactor logging in the parsing engine, in particular the use of expressions like:

```
if parse_context.verbosity > 3:
    parse_match_logging(
        cls.__name__, '_look_ahead_match', 'SC', parse_context=parse_context,
        v_level=cls.v_level,
        bsm=best_simple_match)
```

### PR DESCRIPTION
**Pull request by alanmcruickshank**
This is a bit of a monster. It closes #331.

Importantly it decouples the parser from any responsibility to control it's own output. That means no more passing around verbosity or weird logging routines.

Summary of changes:
- Reorganise tests into subfolders (the was necessary during development to ease testing the new changes).
- Addition of a `Formatter` class which is passed to the `Linter` to which it can delegate all responsibility for formatting as required.
- Removal of the previous `verbosity_logger` so that we now leverage the python logging infrastructure much more, and the configuration of that is handled centrally within `cli.commands` so that anyone importing the `Linter` directly can handle their own logging.
- Refactor of the previous concept of a `ParseContext`. This is now a two level object rather than the previous tuple-like construction. During parsing it handles all the location tracking of where we are and also the appropriate logging of it.

This PR touches a lot of code, so I'd love reviews both of detail but also on the overall approach. I'm not against reworking parts of this if people can see a better approach that's worth taking while I'm in this part of the codebase.

### CODE REVIEW

**Review by pwildenhain — COMMENTED** [review]
💖 Love the decoupling of logging from the `Linter`, and giving the user more control over how `sqlfluff` logging works.

Would you recommend configuring logging when developing the package locally? To help better understand the code flow? If so, I think that should be documented in `CONTRIBUTING.md` as well.

Not giving explicit approval since there was plenty here that went over my head 😵 but I followed the general ideas, and I'm a fan 🙌

File: CHANGELOG.md [hunk 1]
```diff
@@ -13,6 +13,10 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 
 ### Changed
 
+- Big refactor of logging internally. `Linter` is now decoupled from
+  logging so that it can be imported directly by subprojects without
+  needing to worry about wierd output or without the log handing getting
+  in the way of your project.
```
[the remark this answers is not in the document]
Comment (alanmcruickshank) [reply CHANGELOG.md:19]: Great idea. We should probably deal with that separately to here though. Can you add that as a new issue?

**Comment by alanmcruickshank** [comment]
> Would you recommend configuring logging when developing the package locally?

No I don't think so. When I'm developing locally I do use the verbosity option *a lot* though. Quite often I'll use `-vvvvv` to get the deepest logging to work out why something isn't working as intended.  The idea of using `-v/--verbose` together with then new `--logger` option as standard dev practice seem like a good idea. In many ways, the deeper levels of logging are really only useful for debugging rather than any actual direct use.

**Comment by alanmcruickshank** [comment]
I was worried that these changes would involve a performance hit. Looking at the `bench` step in CI, it's actually a performance *gain* on the benchmarks 🙌 .

Oddly though, it increases the time it takes to run the normal test suite, which is puzzling. I'm not sure why. 

Will do a brief investigation before merging.

**Comment by alanmcruickshank** [comment]
Ok, the command that is taking longer is the one where we do lots of logging. I think by adding the capability to defer the work of logging until later (and therefore get the performance gains when *not* logging), we've just added some overhead that means that when we *are* doing all the logging we're doing all the work we were before, *plus* the overhead.

I'm going to call it a day on this and merge unless there is any other feedback from the team?

**Pull request merged**